internal/safari_webdriver.py

```diff
@@ -0,0 +1,1144 @@
+        logging.debug("Collecting page-level metrics")
+        page_data = self.run_js_file('page_data.js')
+        if page_data is not None:
+            task['page_data'].update(page_data)
+        if 'customMetrics' in self.job:
+            custom_metrics = {}
+            requests = None
+            bodies = None
+            for name in self.job['customMetrics']:
+                logging.debug("Collecting custom metric %s", name)
+                custom_script = unicode(self.job['customMetrics'][name])
+                if custom_script.find('$WPT_REQUESTS') >= 0:
+                    if requests is None:
+                        requests = self.get_sorted_requests_json(False)
+                    try:
+                        custom_script = custom_script.replace('$WPT_REQUESTS', requests)
+                    except Exception:
+                        logging.error('Error substituting request data into custom script')
+                if custom_script.find('$WPT_BODIES') >= 0:
+                    if bodies is None:
+                        bodies = self.get_sorted_requests_json(True)
+                    try:
+                        custom_script = custom_script.replace('$WPT_BODIES', bodies)
+                    except Exception:
+                        logging.error('Error substituting request data with bodies into custom script')
+                script = 'var wptCustomMetric = function() {' + custom_script + '};try{return wptCustomMetric();}catch(e){};'
+                try:
+                    custom_metrics[name] = self.execute_js(script)
+                    if custom_metrics[name] is not None:
+                        logging.debug(custom_metrics[name])
+                except Exception:
+                    logging.error('Error collecting custom metrics')
+            path = os.path.join(task['dir'], task['prefix'] + '_metrics.json.gz')
+            with gzip.open(path, GZIP_TEXT, 7) as outfile:
+                outfile.write(json.dumps(custom_metrics))
+        if 'heroElementTimes' in self.job and self.job['heroElementTimes']:
+            hero_elements = None
+            custom_hero_selectors = {}
+            if 'heroElements' in self.job:
+                custom_hero_selectors = self.job['heroElements']
+            logging.debug('Collecting hero element positions')
+            with io.open(os.path.join(self.script_dir, 'hero_elements.js'), 'r', encoding='utf-8') as script_file:
+                hero_elements_script = script_file.read()
+            script = hero_elements_script + '(' + json.dumps(custom_hero_selectors) + ')'
+            hero_elements = self.execute_js(script)
+            if hero_elements is not None:
+                path = os.path.join(task['dir'], task['prefix'] + '_hero_elements.json.gz')
+                with gzip.open(path, GZIP_TEXT, 7) as outfile:
+                    outfile.write(json.dumps(hero_elements))
+
+    def process_message(self, message):
```

Comment:
Is all of the message handling identical to the iOS agent?  If so it would be awesome if we could refactory them both into a shared parent class that both use (or for now add a todo comment to refactor)

Reply:
I do not think we are using same logic. I recall iOS was using web inspector protocol. I will have another look and see if there are logic that could be merged, module being reduced, maybe some refractory work could be done.